pkg/smokescreen/conntrack/conn_tracker_test.go

```diff
@@ -49,5 +49,79 @@ func NewTestTracker(idle time.Duration) *Tracker {
 	sd := atomic.Value{}
 	sd.Store(false)
 
-	return NewTracker(idle, &statsd.NoOpClient{}, logrus.New(), sd)
+	return NewTracker(idle, &statsd.NoOpClient{}, logrus.New(), sd, nil)
+}
+
+// TestConnSuccessRateTracker tests that a ConnTracker with a ConnSuccessRateTracker correctly
+// records connection attempts, calculates a success rate, and expires stored attempts.
+func TestConnSuccessRateTracker(t *testing.T) {
+	type record struct {
+		host    string
+		success bool
+	}
+
+	var testCases = []struct {
+		name         string
+		additions    []record
+		waitTime     time.Duration
+		expectedRate float64
+		totalConns   int
+	}{
+		{"fifty-percent-success", []record{{"foo.com", true}, {"bar.com", false}}, 1 * time.Second, 50.0, 2},
+		{"no-records", []record{}, 0 * time.Second, 100.0, 0},
+		// If we wait beyond the calculation window, we should have no records and return the default 100% rate
+		{"expire-records", []record{{"foo.com", true}, {"bar.com", false}}, 3 * time.Second, 100.0, 0},
+		// Only the most recent record for a host is used in the computation
+		{"dedup-hosts", []record{{"foo.com", false}, {"bar.com", false}, {"foo.com", true}}, 1 * time.Second, 50.0, 2},
```

Comment:
Now that you've added hostname normalization, consider adding a test case that expects `one.foo.com` and `two.foo.com` (and maybe variations with port numbers) to all collapse to `foo.com`.